### DISCUSSION TITLE
Potential path for native indentation support in tree-sitter

### DISCUSSION THREAD

**Discussion by alaviss** [tree-sitter/tree-sitter#2861] [Ideas]
[Principled Parsing for Indentation-Sensitive Languages: Revisiting Landin's Offside Rule](https://michaeldadams.org/papers/layout_parsing/) \[[pdf](https://michaeldadams.org/papers/layout_parsing/LayoutParsing.pdf)] seems to be an interesting paper on how to implement layout rules on top of an LR(k) or GLR grammar. The paper claims that it could describe pretty much all indentation-sensitive grammars, with examples for Haskell and Python.

This appears to work in practice as it was successfully [implemented](https://michaeldadams.org/papers/layout_parsing_2/) for parsec a year later.

**Comment by smheidrich** [comment]
Related issue with prior discussion regarding builtin support for indentation/whitespace-sensitive languages:
- https://github.com/tree-sitter/tree-sitter/issues/219

Someone links this very thread at the end as a potential way to address the concerns raised there regarding what prevents builtin support.

(I'm just mentioning it here to make the link "bidirectional", because apparently GitHub doesn't (yet?) have the "mentioned in" functionality between Issues and Discussions....)
